### PR TITLE
Add size index properties

### PR DESCRIPTION
**Pull request by huachaohuang**
Collect size properties from default cf.
Collect MVCC properties and size properties from write cf.

### CODE REVIEW

File: src/util/properties.rs [hunk 1]
```diff
@@ -87,115 +101,228 @@ impl UserProperties {
...
-    fn decode_u64(&self, k: &str) -> Result<u64, codec::Error>;
+#[derive(Clone, Debug, Default)]
+pub struct IndexHandle {
+    pub size: u64, // The size of the stored block
+    pub offset: u64, // The offset of the block in the file
+}
+
+#[derive(Default)]
+pub struct SizeIndexProperties {
+    pub handles: BTreeMap<Vec<u8>, IndexHandle>,
```
Comment (siddontang) [inline src/util/properties.rs]: do we need to contain total size?

File: src/util/properties.rs [hunk 2]
```diff
@@ -34,40 +43,46 @@ const PROP_NUM_ROWS: &'static str = "tikv.num_rows";
...
 const PROP_NUM_VERSIONS: &'static str = "tikv.num_versions";
 const PROP_MAX_ROW_VERSIONS: &'static str = "tikv.max_row_versions";
-const PROP_NUM_ERRORS: &'static str = "tikv.num_errors";
+const PROP_SIZE_INDEX: &'static str = "tikv.size_index";
 
-#[derive(Clone, Debug, Default)]
+const PROP_SIZE_INDEX_DISTANCE: u64 = 4 * 1024 * 1024;
+
+#[derive(Default)]
 pub struct UserProperties {
```
Comment (zhangjinpeng1987) [inline src/util/properties.rs]: I don't think we need a `UserProperties` to combine several properties collectors, just add them to ColumnFamilyOptions one by one.

File: src/storage/mvcc/reader.rs [hunk 3]
```diff
@@ -514,11 +512,34 @@ impl<'a> MvccReader<'a> {
...
+        }
+        // Aggregate MVCC properties.
+        let mut props = MvccProperties::new();
+        for (_, v) in &*collection {
+            let mvcc = match MvccProperties::decode(v.user_collected_properties()) {
+                Ok(v) => v,
+                Err(_) => return None,
+            };
+            // Filter out properties after safe_point.
+            if mvcc.min_ts > safe_point {
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/reader.rs]: We may reuse mvcc_properties to check how many rows in a region in the future, I think it is better to extract a filter for the aggregator.

**Comment by huachaohuang** [comment]
This PR will be separated into other PRs, close it now.

**Pull request closed without merging**